### ISSUE TITLE
TeamsGroupPolicyAssignment Exported configuration cannot be used because of duplicated key value

### ISSUE THREAD

**Issue by sandrola** [microsoft/Microsoft365DSC#3054]
The exported configuration of TeamsGroupPolicyAssignment cannot be used for dirftdetection because of the key attribute "rank" used by DSC.
#### Details of the scenario you tried and the problem that is occurring
Every Grouppolicyassignment starts with Rank 1, If there are multiple groups, also multiple ranks "1" are present. This exported configuration cannot be compiled because of the Key "Rank".

#### Suggested solution to the issue
As Get-CsGroupPolicyAssignment doesen't generate unique Key in the case above, I suges to Add a unique Key in the Export, and use is as "key" in DSC. (Guid or incremet digit). This was proposed but discarded at the initial Release. We don't thinked about the drift detection. If this is ok, I would implement the solution.

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
# insert configuration here
Node localhost
    {
        TeamsGroupPolicyAssignment 86388c59-d8a1-42d7-b768-cfee48d9c406
        {
            ApplicationId         = $ConfigurationData.NonNodeData.ApplicationId;
            CertificateThumbprint = $ConfigurationData.NonNodeData.CertificateThumbprint;
            Ensure                = "Present";
            GroupDisplayName      = "secgroup3";
            GroupId               = "ff49d90d-ce08-4f5f-92a3-db35816c3ad0";
            PolicyName            = "FirstLineWorker";
            PolicyType            = "TeamsAppSetupPolicy";
            Priority              = 1;
            TenantId              = $ConfigurationData.NonNodeData.TenantId;
        }
        TeamsGroupPolicyAssignment 13fbf8b0-e132-4e83-a3cf-998655571b2f
        {
            ApplicationId         = $ConfigurationData.NonNodeData.ApplicationId;
            CertificateThumbprint = $ConfigurationData.NonNodeData.CertificateThumbprint;
            Ensure                = "Present";
            GroupDisplayName      = "SecGroup2";
            GroupId               = "6eb76881-f56f-470f-be0d-672145d3dcb1";
            PolicyName            = "AllowCalling";
            PolicyType            = "TeamsCallingPolicy";
            Priority              = 1;
            TenantId              = $ConfigurationData.NonNodeData.TenantId;
        }
        TeamsGroupPolicyAssignment f2d379c6-8078-42d9-a949-0bd735b790a2
        {
            ApplicationId         = $ConfigurationData.NonNodeData.ApplicationId;
            CertificateThumbprint = $ConfigurationData.NonNodeData.CertificateThumbprint;
            Ensure                = "Present";
            GroupDisplayName      = "secgroup";
            GroupId               = "4ecd31fa-5da1-4003-a3c8-da7d86edc674";
            PolicyName            = "AllowCallingPreventForwardingtoPhone";
            PolicyType            = "TeamsCallingPolicy";
            Priority              = 2;
            TenantId              = $ConfigurationData.NonNodeData.TenantId;
        }
        TeamsGroupPolicyAssignment 0ce64565-5c3c-4dd3-8b4e-f938f9406ed4
        {
            ApplicationId         = $ConfigurationData.NonNodeData.ApplicationId;
            CertificateThumbprint = $ConfigurationData.NonNodeData.CertificateThumbprint;
            Ensure                = "Present";
            GroupDisplayName      = "secgroup";
            GroupId               = "4ecd31fa-5da1-4003-a3c8-da7d86edc674";
            PolicyName            = "AllOn";
            PolicyType            = "TeamsMeetingPolicy";
            Priority              = 1;
            TenantId              = $ConfigurationData.NonNodeData.TenantId;
        }
    }

#### The operating system the target node is running
OsName               : Microsoft Windows 10 Pro
OsOperatingSystemSKU : 48
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 19041.1.amd64fre.vb_release.191206-1406
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### Version of the DSC module that was used ('dev' if using current dev branch)
1.23.315.1

**Comment by andikrueger** [comment]
This is related to #2006.

**Comment by NikCharlebois** [comment]
Thanks Sandro, out of curiosity, why are you folks naming two different policies the same? I am trying to understand what the business need behind having multiple items with the same name is. Thanks

**Comment by sandrola** [comment]
There is no name for this, the only information are Groupid, Policyname, Policytype, Rank. All attributes can appear doubled.

**Comment by sandrola** [comment]
You are right, but the groupid as key is not the solution, the same group can be used for different policy/policy types. See this output:

Get-CsGroupPolicyAssignment | fl *

CreatedBy   : d37cec8b-ad61-4a2f-a424-c6f59428ac88
CreatedTime : 3/15/2023 11:36:58 AM
GroupId     : ff49d90d-ce08-4f5f-92a3-db35816c3ad0
PolicyName  : FirstLineWorker
PolicyType  : TeamsAppSetupPolicy
Priority    : 1

CreatedBy   : d37cec8b-ad61-4a2f-a424-c6f59428ac88
CreatedTime : 1/6/2023 2:46:54 PM
GroupId     : 6eb76881-f56f-470f-be0d-672145d3dcb1
PolicyName  : AllowCalling
PolicyType  : TeamsCallingPolicy
Priority    : 1

CreatedBy   : d37cec8b-ad61-4a2f-a424-c6f59428ac88
CreatedTime : 1/3/2023 4:15:39 PM
GroupId     : 4ecd31fa-5da1-4003-a3c8-da7d86edc674
PolicyName  : AllowCallingPreventForwardingtoPhone
PolicyType  : TeamsCallingPolicy
Priority    : 2

CreatedBy   : d37cec8b-ad61-4a2f-a424-c6f59428ac88
CreatedTime : 1/3/2023 1:11:56 PM
GroupId     : 4ecd31fa-5da1-4003-a3c8-da7d86edc674
PolicyName  : AllOn
PolicyType  : TeamsMeetingPolicy
Priority    : 1

**Comment by andikrueger** [comment]
I see. the difference. We need to make PolicyType a Key as well. Then it will work.

**Comment by sandrola** [comment]
What would solve the problem is to introduce an "Id" (numeric incremental generated in export) wich is used only by DSC for compiling. This Id would be the Key.

**Comment by andikrueger** [comment]
We do have this discussion in #2006 and are looking for a way to get around it.

**Comment by andikrueger** [comment]
Just to be a bit more specific as in my previous answer: 

We need a solution that would fit either use case: 

1. Creating a configuration from scratch
2. Creating a configuration as result from an export

In Export we could easily add any number to the resource. It would still require the resource to have an additional key that would only be used for the compilation and uniqueness of a resource. As soon as someone would create an export, we would not be aware of the original ID to make a meaningful comparison of both configurations. Even in the case of Test-TargetResource, this could cause troubles.

**Comment by ykuijs** [comment]
The code first checks if the specified GroupId exists and when it does not, uses the DisplayName to retrieve the group based on name. This allowed the export to be applied to a different tenant, where the group has a different ID but the same name.

In the Intune resource, we have the same issue. There we specified the Id as key parameter and the DisplayName as required. That way we always have the needed information at available. I would suggest that we do the same here. Maybe not the ideal solution, but the best we have available now and can implement before Wednesdays release.

@andikrueger Can you please update you PR, update it with Dev and set it to final. Then I can merge it into Dev and include this in our upcoming release.

**Comment by andikrueger** [comment]
@ykuijs Just set the PR to active. Looks like there are no conflicts... Will update the code to add the GroupID check.
